smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.endpoints
+
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
+import software.amazon.smithy.kotlin.codegen.core.withBlock
+import software.amazon.smithy.kotlin.codegen.lang.KotlinTypes
+import software.amazon.smithy.kotlin.codegen.model.boxed
+import software.amazon.smithy.kotlin.codegen.utils.doubleQuote
+import software.amazon.smithy.kotlin.codegen.utils.getOrNull
+import software.amazon.smithy.rulesengine.language.EndpointRuleset
+import software.amazon.smithy.rulesengine.language.eval.Value
+import software.amazon.smithy.rulesengine.language.lang.Identifier
+import software.amazon.smithy.rulesengine.language.lang.parameters.Parameter
+import software.amazon.smithy.rulesengine.language.lang.parameters.ParameterType
+
+private const val DEFAULT_DEPRECATED_MESSAGE =
+    "This field is deprecated and no longer recommended for use."
+
+private const val CLASS_NAME = "EndpointParameters"
+
+/**
+ * Renders the struct of parameters to be passed to the endpoint provider for resolution.
+ */
+class EndpointParametersGenerator(
+    private val writer: KotlinWriter,
+    rules: EndpointRuleset,
+) {
+    private val params: List<KotlinEndpointParameter> = rules.parameters.toList().map {
+        KotlinEndpointParameter(
+            it.name.toKotlin(),
+            it.type.toSymbol(),
+            it.isRequired,
+            it.defaultValue.getOrNull()?.toKotlinLiteral() ?: "null",
+            it.documentation.getOrNull(),
+            it.deprecated.getOrNull(),
+        )
+    }
+
+    fun render() {
+        renderDocumentation()
+        writer.withBlock("public class #L private constructor(builder: Builder) {", "}", CLASS_NAME) {
+            renderFields()
+            renderCompanionObject()
+            write("")
+            if (params.any(KotlinEndpointParameter::isRequired)) {
+                renderInit()
+                write("")
+            }
+            renderEquals()
+            write("")
+            renderHashCode()
+            write("")
+            renderToString()
+            write("")
+            renderCopy()
+            write("")
+            renderBuilder()
+        }
+    }
+
+    private fun renderFields() {
+        params.forEach {
```

Comment:
**Question**: In `StructureGenerator`, we sort members before rendering them—not only at field declaration but in `equals`, `hashCode`, etc. This generally makes classes with a lot of members easier to navigate and parse. Should we consider doing the same here?